### DISCUSSION TITLE
Link types/custom handlers for links

### DISCUSSION THREAD

**Discussion by joshpetit** [nvim-neorg/neorg#961] [Ideas]
Org mode has this feature I really like where you're able to define your own link handler/type. It [looks like this](https://youtu.be/mRGFE-Pn86Q?t=70) in practice and the [docs are here](https://orgmode.org/manual/Adding-Hyperlink-Types.html#Adding-Hyperlink-Types).

I think it would be pretty cool if neorg could support something like this. For instance, say:
```org
{man:printf}[printf manual page]
```
And then when people click on the link, they could configure neorg to call a function as such:
````lua
--- "context" maybe having info like the file, workspace, buffer, line+column etc of the link that was called. 
local man_link_handler = function(context, link)
   vim.fn.execute('!man ' .. link)
end

````

In addition to that, org mode lets you define how it is exported as well. So that would look something like:

```lua
local man_link_md_exporter = function(context, link)
   return 'https://manualpage/' .. link
end
````

Implementation wise, I've read that neorg broadcasts events, so I imagine this wouldn't be too too difficult to implement if there is a "linked clicked" event and we can specify which handlers can listen to that event.

One thing I wish orgmode had with this was the ability to set explicit parameters to the links (it's relatively implicit since I would make parameters by just butting spaces between what I consider params and splitting there). So if something like
```org
{man: exe="/usr/bin/man", query="printf", fallbackpaths=["/bin/man", "/opt/man"]}[Printf man page]
```
which would then be passed as a table to `context.params` would be  pretty powerful.

**Comment by SevorisDoe** [comment]
I *suspect* but am not sure that Neorg‘s macro systems should let you do this using tag extensions and the Janet macro system. The idea of defining a custom export target is definitely in there.

**Reply by joshpetit** [reply]
Thanks! While the exporting is a ways there, there is quite a bit of benefit to being able to run custom functions to figure out the value or action of the link. More complex examples can be seen in the video I linked where when I clicked on a link it gave me the option to copy the bible text from the reference to my clipboard.

Looking at the specification it seems to mostly be templating that gets expanded later. But you can correct me if I'm wrong.

The main thing I'm advocating for though I'd being able to evaluate functions to handle link exporting and handling.

**Reply by SevorisDoe** [reply]
The macro system is a bit unintuitive, but there‘s a set of outcomes when a macro is processed. If the macro returns concrete norg output, it gets "baked", and the macro is replaced by its output. If the macro returns a neorg abstract object, it cannot be baked, and instead you have the macro text with a data structure "behind" it which can do things like set (anti)conceil settings and so on. 

That later mechanism AFAII should allow you to define the data that something like a custom module can link into so that when you‘re calling for link use on the macro-defined custom link, your logic runs and produces the output. 

(I‘ve discussed the tag system and macros before to implement equivalents to Markdown hashtags and Dataview (inline) key-value pairs. The syntax is definitely designed to specify a function and a set of arguments, so you should be able to encode the data structure you want without breaking semantic parsing. And tags can be attached like link specifications can. ) 

Otherwise, we may have to discuss extending the norg syntax. Having custom functions on links is something I also find interesting.

**Reply by joshpetit** [reply]
Oh ok fantastic, I see [here](https://github.com/nvim-neorg/norg-specs/blob/35a52c4510f3deabfe52112836504ec7f8fa08fd/1.0-semantics.norg#L318) there is section on extendable links but doesn't give context on how to invoke lua code with them.

I'm trying to follow along with the docs (reading through the specs and then making my way through semantics file) but I can't seem to even get simple macros to expand. I also don't see reference to how functions can be called. I think I may just not grok how any of this works to be honest. I'll have to possibly try another week to sit down and try to understand the system.

My preference would be for something very simple (as I outlined in my original idea), but I am open to trying to learn if it offers significant advantages over what I outlined.

**Reply by SevorisDoe** [reply]
@vhyrro this is really into your domain at this point.
